File: app/src/main/java/ch/bailu/aat/services/beacon/BeaconService.java
package ch.bailu.aat.services.beacon;

import java.io.IOException;
import java.net.SocketException;
import java.net.InetSocketAddress;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.AsyncTask;

import beacon.BeaconClient;

import ch.bailu.aat.gpx.GpxInformation;
import ch.bailu.aat.util.AppBroadcaster;
import ch.bailu.aat.util.WithStatusText;
import ch.bailu.aat.services.ServiceContext;
import ch.bailu.aat.services.VirtualService;
import ch.bailu.aat.services.location.LocationService;

/**
 * Integrate the #BeaconClient into AAT as a #VirtualService.
 */
public final class BeaconService extends VirtualService implements WithStatusText {
    private BeaconClient client;
    private GpxInformation oldLocation = GpxInformation.NULL;

    private AsyncTask<GpxInformation, Integer, Integer> submitTask;

    public BeaconService(ServiceContext sc) {
        super(sc);

        try {
            // TODO: no hard-coded IP address
            client = new BeaconClient(new InetSocketAddress("138.201.185.127", 5598), 42);
        } catch (SocketException e) {
            return;
        }

        AppBroadcaster.register(getContext(), onLocation,
                                AppBroadcaster.LOCATION_CHANGED);
    }

    class SubmitTask extends AsyncTask<GpxInformation, Integer, Integer> {
        @Override
        protected Integer doInBackground(GpxInformation... i) {
            try {
                client.sendFix(i[0].getLatitude(), i[0].getLongitude());
            } catch (IOException e) {
            }

            return 0;
        }

        @Override
        protected void onPostExecute(Integer result) {
            submitTask = null;
        }
    }

    public void close() {
        if (submitTask != null)
            submitTask.cancel(true);

        if (client != null)
            client.close();

        getContext().unregisterReceiver(onLocation);
    }

    private void submit(GpxInformation location) {
        if (client == null)
            return;

        if (submitTask != null) {
            submitTask.cancel(true);
            submitTask = null;
        }

        submitTask = new SubmitTask().execute(location);
    }

    private void submit() {
        final LocationService l = getSContext().getLocationService();
        if (l.hasLoggableLocation(oldLocation)) {
            oldLocation = l.getLoggableLocation();
            submit(oldLocation);
        }
    }

    private final BroadcastReceiver onLocation = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            submit();
        }
    };

    @Override
    public void appendStatusText(StringBuilder builder) {
        builder.append("<p>Beacon!</p>");
    }

    public GpxInformation getLoggerInformation() {
        return oldLocation;
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/service/beacon/BeaconService.kt
package ch.bailu.aat_lib.service.beacon

import beacon.BeaconClient
import ch.bailu.aat_lib.app.AppContext
import ch.bailu.aat_lib.logger.AppLog
import ch.bailu.aat_lib.broadcaster.AppBroadcaster
import ch.bailu.aat_lib.broadcaster.BroadcastReceiver
import ch.bailu.aat_lib.broadcaster.Broadcaster
import ch.bailu.aat_lib.gpx.information.GpxInformation
import ch.bailu.aat_lib.preferences.OnPreferencesChanged
import ch.bailu.aat_lib.preferences.StorageInterface
import ch.bailu.aat_lib.preferences.beacon.SolidBeaconEnabled
import ch.bailu.aat_lib.preferences.beacon.SolidBeaconServer
import ch.bailu.aat_lib.preferences.beacon.SolidBeaconKey
import ch.bailu.aat_lib.service.ServicesInterface
import ch.bailu.aat_lib.service.VirtualService
import ch.bailu.aat_lib.service.background.BackgroundTask
import ch.bailu.aat_lib.util.WithStatusText
import java.io.IOException
import java.net.InetSocketAddress
import java.net.SocketException

/**
 * Integrate the #BeaconClient into AAT as a #VirtualService.
 */
class BeaconService(
    private val services: ServicesInterface,
    private val broadcaster: Broadcaster,
    private val storage: StorageInterface
) : VirtualService(), WithStatusText, BeaconServiceInterface, OnPreferencesChanged {
    private var client: BeaconClient? = null
    private var location: GpxInformation = GpxInformation.NULL

    private val onLocation = BroadcastReceiver { submit() }

    init {
        broadcaster.register(AppBroadcaster.LOCATION_CHANGED, onLocation)
        storage.register(this)
        maybeCreateClient(storage)
    }

    @Synchronized
    override fun close() {
        storage.unregister(this)
        broadcaster.unregister(onLocation)
        client?.close()
        client = null
    }

    @Synchronized
    private fun maybeCreateClient(storage: StorageInterface) {
        client?.close()
        client = null

        if (!SolidBeaconEnabled(storage).isEnabled) {
            AppLog.d(this, "Beacon is disabled")
            return
        }

        val server = SolidBeaconServer(storage).getValue()
        val key = SolidBeaconKey(storage).getValue()

        if (server == null || key.toLong() == 0L) {
            AppLog.w(this, "Beacon configuration incomplete")
            return
        }

        try {
            client = BeaconClient(server, key.toLong())
            AppLog.i(this, "Beacon initialized, submitting to " + server)
        } catch (e: SocketException) {
            AppLog.e(this, e, "Beacon initialization failed")
        }
    }

    override fun onPreferencesChanged(storage: StorageInterface, key: String) {
        if (key.startsWith("BEACON_")) {
            AppLog.d(this, "Beacon preferences changed")
            maybeCreateClient(storage)
        }
    }

    private fun submit(gpxLocation: GpxInformation) {
        client?.let { beaconClient ->
            services.getBackgroundService().process(BeaconSubmitTask(beaconClient, gpxLocation))
        }
    }

    private fun submit() {
        val locationService = services.getLocationService()
        val newLocation = locationService.getLoggableLocationOrNull(location)
        if (newLocation != null) {
            location = newLocation
            submit(location)
        }
    }

    private class BeaconSubmitTask(
        private val beaconClient: BeaconClient,
        private val gpxLocation: GpxInformation
    ) : BackgroundTask() {
        override fun bgOnProcess(appContext: AppContext): Long {
            try {
                beaconClient.sendFix(gpxLocation.getLatitude(), gpxLocation.getLongitude())
            } catch (e: IOException) {
                // Ignore IOException - beacon submission failed
            }
            return 0 // No size to report for network operations
        }
    }

    override fun appendStatusText(builder: StringBuilder) {
        builder.append("<p>Beacon!</p>")
    }

    override fun getLoggerInformation(): GpxInformation {
        return location
    }
}
